Internals Bugfixes & Internals-Compatible Helmets

Floofies opened the pull request:
## About The Pull Request
This PR contains a modest refactor, 1 major and 2 minor feature additions, and bugfixes, for several portions of the code pertinent to internals / mob-equipped air tanks.

The major additional feature of this PR adds internals support to many sealed helmets by adding a new `HEADINTERNALS` flag to them. The minor additions are AltClick on internals tanks, and an "externals" system for @Fikou

### Additions

#### Internals-Compatible Helmets

There are several careful considerations behind the changes to helmets, such as if a helmet can theoretically connect to an air tank based on its physical design. Specifically, I added internals support to helmets which at least fully enclose the user's head, and which satisfy at least one of my following criteria:

- _Criterion A_. The helmet must be hermetically sealed, i.e. it must at least have the `STOPSPRESSUREDAMAGE` flag. This includes all spacesuit helmets, space-worthy MOD suit helmets, the changeling's fleshmass, the hardened cult hoodie, and the atmospherics helmet.
- _Criterion B_. The helmet may not be fully sealed but must block significant airflow, i.e. it must have the `BLOCK_GAS_SMOKE_EFFECT` and `PEPPERPROOF` flags. This includes the biohood.

Additionally, in order to get MOD suit helmets to toggle internals as expected, I hooked into the `mod/control/proc/seal_part` proc to check internals status.

#### External Internals Tanks

I was also asked to add an "externals" system to carbons which allows them to breathe from an air tank which they have not equipped. One of the later first uses for such an externals hook will perhaps be a MOD suit module.

#### AltClick on Internals Tanks

I have added the AltClick proc to internals tanks `/obj/item/tank/internals` which allows humans to toggle their internals quickly.

### BugFixes & Internals Refactor

My original intention was only to add internals compatibility to helmets, however I encountered several bugs in the related code which this PR includes fixes for. The bug fixes included in this PR correct the following issues:

1. When a human adjusts any mask, its `wear_mask_update` function and `update_action_buttons_icon` procs are called in inappropriate conditions, i.e. when the mask was not equipped by that mob.
2. The Alt+click functionality of masks isn't working, due to a typo (stray `!`) in the `mask/breath/AltClick` proc.
3. When opening internals, any worn mask is adjusted over the face regardless of whether or not it's compatible with internals, due to `tank/toggle_internals` not checking for the `MASKINTERNALS` flag first.
4. When a human gives their open internals tank to another mob via the stripping menu, the tank can be opened for both mobs at once for a limited but noticeable period of time.

To fix the aforementioned bugs and easily add internals support to helmets, I quickly realized that I had to address a code health issue first. More concisely, I realized that I had the opportunity to improve relevant critical sections of code which interoperated with the code I was working on. To that end, I added several ubiquitous utility procs which allowed me to make the code a lot more DRY, eliminate repetitive and unreliable conditionals, separate some concerns between `tank` and `mob`, and ultimately better centralize the logic for toggling internals.

I wound up porting internals functionality entirely out of `/obj/item/tank` and into `/mob/living/carbon`. Several utility procs for working with internals were also added to `mob/living/carbon` and `mob/living/carbon/human`, allowing carbons to manage their own internals via `open_internals(tank)` and `can_breathe_internals()`etc.

To fix bugs which allowed multiple mobs to breathe from the same tank, the following procs were also added to `obj/item/tank`:

- `equipped(mob/living/user, slot, initial)` - Closes the open air tank if it was equipped by a mob other than the current user.
- `dropped(mob/living/user, silent)` - Closes the open air tank if it got dropped by it's current user.
- `after_internals_opened(mob/living/carbon/human/human_target)` - Called by carbons after they connect the tank to their breathing apparatus.
- `after_internals_closed(mob/living/carbon/human/human_target)` - Called by carbons after they disconnect the tank from their breathing apparatus.

## Why It's Good For The Game
This PR introduces a feature which allows several sealed helmets, such as spacesuit helmets, to be used as breathing apparatuses. Where appropriate, this PR adds internals compatibility to spacesuit helmets, space-worthy MOD suit helmets, the biohood, changeling fleshmass, hardened cult hoodie, and atmos helmet. I think most people would agree that in-general a space suit _should be_ capable of handling the connection to an internal air supply, and that they can easily do so without an extra mask attached directly to the face. This PR also includes a refactor to internals code in order to improve overall code health and fix several bugs.

The most noticeable bug fixed was that the Alt-click adjustment function of breath/gas masks was non-functional. I took the liberty of bringing the same AltClick functionality to internals tanks as well, allowing you to toggle your internals on quickly!

In my opinion, my changes make the game a bit more realistic and fun. I think that my refactor also definitely makes the code more maintainable. The refactor decouples some shared logic between air tank functions and human/carbon functions, separating those concerns more appropriately and making the system more bug-resistant in the future.

## Changelog

:cl: A.C.M.O.
add: Hermetically sealed and space-worthy helmets can now be used as breathing apparatuses!
add: Internals air tanks can now be toggled via AltClick.
fix: Breath masks are fixed, and can now be toggled via AltClick.
refactor: Refactored and deduplicated a significant portion of internal air tank functions.
fix: Fixed a bug that caused any type of equipped mask to be auto-adjusted when toggling internals.
fix: Fixed a bug that prevented toggling of internals via the stripping menu if the mob only had a breathing tube.
fix: Fixed a bug that allowed two mobs to breathe from the same air tank after exchanging it between them.
/:cl:

Floofies committed:
Core internals refactor. Add open/close_internals procs to /obj/item/tank. Add open/close_internals, can_breathe_internals, and invalid_internals procs to /mob/living/carbon. Fixed bug in human_stripping.dm preventing changing internals for breathing tube users. Fixed bug in /obj/item/tank/toggle_internals that was auto-adjusting non-internals masks. Fixed bug in slements/strippable that caused two mobs to breathe from the same internals.

Floofies committed:
Fixed bug in mask/proc/adjustmask that unintentially called other mobs wear_mask_update and updated their icons. Fixed bug in mask/breath/AltClick that prevented mobs adjusting breathmasks.

Floofies committed:
Add support to internals for helmets with MASKINTERNALS flag

Floofies committed:
Add internals support to helmets: Added MASKINTERNALS flags to appropriate MOD and space helmets, biohood, changeling fleshmass, hardened cult hoodie, and atmos helmet.

Floofies committed:
Removed redundant conditional, replaced with better docs

Floofies committed:
Fixed bug in hand/finish_equip that caused two mobs to breathe from the same internals.

san7890 commented on code/game/objects/items/tanks/tanks.dm:
```diff
@@ -48,33 +48,53 @@
-/obj/item/tank/proc/toggle_internals(mob/user)
-	var/mob/living/carbon/human/H = user
-	if(!istype(H))
+/obj/item/tank/proc/open_internals(mob/living/carbon/human/human_target)
+	if(human_target.internal)
+		to_chat(human_target, span_notice("You switch your internals to [src]."))
+	else
+		to_chat(human_target, span_notice("You open [src] valve."))
+	human_target.open_internals(src)
+/obj/item/tank/proc/close_internals(mob/living/carbon/human/human_target)
+	to_chat(human_target, span_notice("You close [src] valve."))
+	human_target.close_internals()
+
+/obj/item/tank/proc/toggle_internals(mob/mob_target)
```
DMDoc what these procs do, you can read more about that here: (https://github.com/tgstation/tgstation/blob/master/.github/guides/AUTODOC.md)

Floofies replied:
Fixed by commit c7d5cf61ea922486cdc0586fa494b3c339926c96

san7890 commented on code/modules/mob/living/carbon/inventory.dm:
```diff
@@ -174,6 +174,33 @@
 
 	update_equipment_speed_mods()
 
+// Returns TRUE if an air tank compatible helmet is equipped.
+/mob/living/carbon/proc/can_breathe_helmet()
+	return head && isclothing(head) && (head.clothing_flags & MASKINTERNALS) ? TRUE : FALSE
+// Returns TRUE if an air tank compatible mask is equipped.
+/mob/living/carbon/proc/can_breathe_mask()
+	return wear_mask && isclothing(wear_mask) && (wear_mask.clothing_flags & MASKINTERNALS) ? TRUE : FALSE
+// Returns TRUE if a breathing tube is equipped.
+/mob/living/carbon/proc/can_breathe_tube()
+	return getorganslot(ORGAN_SLOT_BREATHING_TUBE) ? TRUE : FALSE
+// Returns TRUE if an air tank compatible mask or breathing tube is equipped.
+/mob/living/carbon/proc/can_breathe_internals()
+	return can_breathe_tube() || can_breathe_mask() || can_breathe_helmet()
+// Returns TRUE if air tank is open and mob lacks apparatus, or if the tank moved away from the mob.
+/mob/living/carbon/proc/invalid_internals()
+	return internal && (internal.loc != src || !can_breathe_internals()) ? TRUE : FALSE
```
```suggestion
/// Returns TRUE if an air tank compatible helmet is equipped.
/mob/living/carbon/proc/can_breathe_helmet()
	return head && isclothing(head) && (head.clothing_flags & MASKINTERNALS) ? TRUE : FALSE

/// Returns TRUE if an air tank compatible mask is equipped.
/mob/living/carbon/proc/can_breathe_mask()
	return wear_mask && isclothing(wear_mask) && (wear_mask.clothing_flags & MASKINTERNALS) ? TRUE : FALSE

/// Returns TRUE if a breathing tube is equipped.
/mob/living/carbon/proc/can_breathe_tube()
	return getorganslot(ORGAN_SLOT_BREATHING_TUBE) ? TRUE : FALSE

/// Returns TRUE if an air tank compatible mask or breathing tube is equipped.
/mob/living/carbon/proc/can_breathe_internals()
	return can_breathe_tube() || can_breathe_mask() || can_breathe_helmet()

/// Returns TRUE if air tank is open and mob lacks apparatus, or if the tank moved away from the mob.
/mob/living/carbon/proc/invalid_internals()
	return internal && (internal.loc != src || !can_breathe_internals()) ? TRUE : FALSE
```

Just so it docs properly, have `///`. There's two more procs below this that could also use it, but I think this suggestion would get rather lengthy.

Floofies replied:
Thanks! Also fixed more by commit c7d5cf61ea922486cdc0586fa494b3c339926c96

LemonInTheDark commented:
So if I just have a space suit helmet on, that works? doesn't relate to a suit at all

Floofies commented:
@LemonInTheDark Correct, only the helmet is needed, and it can be used as a breathing apparatus without a suit.

LemonInTheDark commented on code/modules/mob/living/carbon/inventory.dm:
```diff
@@ -174,6 +174,33 @@
+	return can_breathe_tube() || can_breathe_mask() || can_breathe_helmet()
+// Returns TRUE if air tank is open and mob lacks apparatus, or if the tank moved away from the mob.
+/mob/living/carbon/proc/invalid_internals()
+	return internal && (internal.loc != src || !can_breathe_internals()) ? TRUE : FALSE
```
Newlines. not a massive fan of splitting these into procs because lmao proccall overhead, but I see why you did it

Floofies replied:
Fixed by commit df5906d634dfaf8a37a1ec83b277c36e43fbfd87

Floofies committed:
Add extra slashes to preserve autodoc

Co-authored-by: san7890 <the@san7890.com>

LemonInTheDark commented on code/modules/mob/living/carbon/inventory.dm:
```diff
@@ -174,6 +174,33 @@
+/mob/living/carbon/proc/can_breathe_internals()
+	return can_breathe_tube() || can_breathe_mask() || can_breathe_helmet()
+// Returns TRUE if air tank is open and mob lacks apparatus, or if the tank moved away from the mob.
+/mob/living/carbon/proc/invalid_internals()
+	return internal && (internal.loc != src || !can_breathe_internals()) ? TRUE : FALSE
```
```suggestion
/mob/living/carbon/proc/invalid_internals()
	return !!(internal && (internal.loc != src || !can_breathe_internals()))
```
that or just make it an if, less confusion that way

Floofies replied:
Hi, I disagree with your change because the ternary is more widely used in the codebase, and for this same purpose. Before using it, I spoke to several people in Discord and made sure to do research on how widely used the ternary was for this purpose. Furthermore, I personally think the double inversion is a superfluous operation when used in this kind of context.

LemonInTheDark replied:
I can find 2 uses of ternaries like this (in that they return true if the case is true, and false otherwise). I don't like this use because it makes the conditional slightly visually muddy.
I can understand wanting to be explicit, but if you're going to be explicit just use if.

It would also be acceptable to just straight up return the string of blocks. I'm aware of the typesafety stuff, but that's an issue over pretty much the whole codebase, and if someone seriously tries to pull values out of the return value I'm gonna hittem over the head with a rock. (It's also niche enough behavior that I doubt anyone who would do it would ever know they could do it)

Floofies replied:
I found 69 different instances of such a ternary in the TG codebase... but ok!

Floofies replied:
Fixed by commit 3d3fc239644f88ab70d29e685536e8afd34f316d

Floofies committed:
Add extra slashes to preserve autodoc

Floofies commented on code/datums/elements/strippable.dm:
```diff
@@ -226,6 +226,11 @@
 
 	var/mob/mob_source = source
+	if (iscarbon(user))
+		var/mob/living/carbon/carbon_user = user
+		// Close an open air tank if it exchanged from user to owner.
+		if (equipping == carbon_user.internal)
+			carbon_user.close_internals()
 	mob_source.equip_to_slot(equipping, item_slot)
```
How do you propose I do this? Currently the `get_breath_from_internal` function calls `internal.remove_air_volume(volume_needed)` which is the only active call-back to the tank while the mob is just wearing it. If I were to implement your proposed change, I could possibly do it inside `remove_air_volume`. However, I disagree with the change as it goes against the law of Demeter and re-introduces the shared concerns I was intending to eliminate.

LemonInTheDark replied:
I don't like how you need to have code like this in every instance of something that "takes" the internals tank. I think that's silly, and hinders maintainability. (had no idea what LoD was but I do see your point)

I had it backwards. carbons should manage their internals. Using a setter for the var, they can register for it moving, and if that happens we can just deactivate them.

Floofies replied:
I decided to split up those two concerns between `connect_internals` etc in carbons, and `open_internals` etc in tanks. It helped me to implement the requested change and also move those assignments to the appropriate `equipped` etc procs.

Floofies replied:
Fixed by commit 3d3fc239644f88ab70d29e685536e8afd34f316d

Floofies replied:
Ok, I have completely ported 99% of the pertinent internals codee from item/tank to living/carbon and carbon/human!

Floofies committed:
Adding missing newlines etc

Floofies commented on code/game/objects/items/tanks/tanks.dm:
```diff
@@ -48,33 +48,53 @@
+	else
+		to_chat(human_target, span_notice("You open [src] valve."))
+	human_target.open_internals(src)
+/obj/item/tank/proc/close_internals(mob/living/carbon/human/human_target)
```
Fixed by commit 989296b49828a640004d0a52b7ed2e3298362235

Floofies commented on code/modules/clothing/suits/bio.dm:
```diff
@@ -6,7 +6,7 @@
 	icon_state = "bio"
 	inhand_icon_state = "bio_hood"
-	clothing_flags = THICKMATERIAL | BLOCK_GAS_SMOKE_EFFECT | SNUG_FIT | PLASMAMAN_HELMET_EXEMPT
+	clothing_flags = THICKMATERIAL | BLOCK_GAS_SMOKE_EFFECT | SNUG_FIT | PLASMAMAN_HELMET_EXEMPT | MASKINTERNALS
```
It is a clothing flag which can be used on more than just masks, however I see why it might be confusing to re-use it. Would you like me to add a `HEADINTERNALS` define instead?

LemonInTheDark replied:
yes

Floofies replied:
Fixed by commit d04a8412f562beec3ec01a88e7c51f305d90c5d5 and commit 3d3fc239644f88ab70d29e685536e8afd34f316d

Floofies committed:
Add define for HEADINTERNALS clothing flag

Floofies committed:
Removed internals conditionals from elements/strippable.dm and carbon/carbon_stripping.dm. Refactored interface between mob and tank, add cutoff_internals proc, add carbon/connect_internals and carbon/disconnect_internals procs and move to_chat out of tank/open_internals and tank/close_internals. Add breathing_mob var and dropped/equipped procs to item/tank. Added HEADINTERNALS clothing flag and replaced MASKINTERNALS with it on headgear. Refactored carbon/can_breathe procs to be more concise.

Floofies commented on code/modules/mob/living/carbon/carbon_stripping.dm:
```diff
@@ -107,7 +107,11 @@
+		var/mob/living/carbon/carbon_user = user
+		// Close an open air tank if it exchanged from user to owner.
+		if (equipping == carbon_user.internal)
+			carbon_user.close_internals()
```
Fixed by commit 3d3fc239644f88ab70d29e685536e8afd34f316d

Floofies committed:
Remove stray whitespace

Floofies committed:
Replaced assigntments to carbon/internal with calls to open_internals and cutoff_internals

Floofies committed:
Fixed typo in comment

Floofies committed:
Moved most internals code to mob/living/carbon, including toggle_internals. Added externals feature at request of maintainer. Moved mod_suit_toggled from carbon/human to item/mod/control and fixed a typo in my RegisterSignal

Floofies committed:
Moving change from remote

Floofies commented:
This PR is looking officially done as long as there aren't any other requests or tweaks to make 👍

Shadowflame909 commented:
Should add to the swat helmet too. I don't know why but lore-wise that's always been a functional space-helmet replacement. Better too since it has armor.

Floofies committed:
Add AltClick toggle to internals tanks, added missing change in /obj/item/tank/ui_data and removed QDEL_NULL from /mob/living/carbon/human/get_status_tab_items. Small bugfixes to prior commits

Floofies commented:
I have debugged and accounted for a couple of edge cases. I also added missing code for the stat panel, and removed a stinky `QDEL_NULL` which was inside `/mob/living/carbon/human/get_status_tab_items`. I think that this PR is as mature as I can make it, but let me know if there's anything that looks odd and I can touch it up.

@Shadowflame909 I hesitate to touch that awkwardly implemented SWAT helmet. From my perspective it does not actually cover the face, and should be worn with a SWAT mask anyways. If the sprite of the helmet actually covered the face, or had a mode in which it did, then I could consider it.

Floofies committed:
Added istype to make adjustmask usable by noncarbon mobs

ShizCalev commented on code/game/objects/items/tanks/tank_types.dm:
```diff
@@ -8,6 +8,16 @@
+/// Allows carbon to toggle internals via AltClick of the equipped tank.
+/obj/item/tank/internals/AltClick(mob/user)
+	..()
+	if((src.loc == user) && user.canUseTopic(src, be_close = TRUE, no_dexterity = FALSE, no_tk = TRUE, need_hands = TRUE))
```
```suggestion
	if((loc == user) && user.canUseTopic(src, be_close = TRUE, no_dexterity = FALSE, no_tk = TRUE, need_hands = TRUE))
```

Floofies replied:
Fixed by commit ffe83e90a18854cc75d31939743d313bb50ad97c

ShizCalev commented on code/modules/mob/living/carbon/human/inventory.dm:
```diff
@@ -287,11 +287,49 @@
+		return TRUE
+	// Invalid helmet and missing mask.
+	// Don't show the "isn't sealed" message for non-helmet headgear, such as hats.
+	if (head && (istype(head, /obj/item/clothing/head/mod) || istype(head, /obj/item/clothing/head/helmet)))
```
```suggestion
	if (istype(head, /obj/item/clothing/head/mod) || istype(head, /obj/item/clothing/head/helmet))
```
istype /is/ the sanity check.

ShizCalev commented on code/modules/clothing/masks/_masks.dm:
```diff
@@ -81,9 +81,14 @@
 		if(adjusted_flags)
 			slot_flags = adjusted_flags
-	if(user)
+	if(!user || !istype(user))
```
```suggestion
	if(!istype(user))
```

Floofies replied:
Fixed by commit ffe83e90a18854cc75d31939743d313bb50ad97c

ShizCalev commented on code/modules/mob/living/carbon/inventory.dm:
```diff
@@ -168,13 +168,157 @@
-//handle stuff to update when a mob equips/unequips a mask.
+/// Returns TRUE if an air tank compatible helmet is equipped.
+/mob/living/carbon/proc/can_breathe_helmet()
+	if (head && isclothing(head) && (head.clothing_flags & HEADINTERNALS))
```
```suggestion
	if(isclothing(head) && (head.clothing_flags & HEADINTERNALS))
```

Floofies replied:
Fixed by commit ffe83e90a18854cc75d31939743d313bb50ad97c

ShizCalev commented on code/modules/mob/living/carbon/inventory.dm:
```diff
@@ -168,13 +168,157 @@
+
+/// Returns TRUE if an air tank compatible mask is equipped.
+/mob/living/carbon/proc/can_breathe_mask()
+	if (wear_mask && isclothing(wear_mask) && (wear_mask.clothing_flags & MASKINTERNALS))
```
```suggestion
	if(isclothing(wear_mask) && (wear_mask.clothing_flags & MASKINTERNALS))
```

Floofies replied:
Fixed by commit ffe83e90a18854cc75d31939743d313bb50ad97c

Floofies committed:
Removing src.loc usage, remove excessive conditionals

Floofies commented on code/game/objects/items/tanks/tanks.dm:
```diff
@@ -44,38 +44,38 @@
+/obj/item/tank/dropped(mob/living/user, silent)
+	. = ..(user, silent)
+	// Close open air tank if it got dropped by it's current user.
+	if (breathing_mob && src.loc != breathing_mob)
```
Fixed by commit ffe83e90a18854cc75d31939743d313bb50ad97c

Floofies commented on code/modules/clothing/masks/_masks.dm:
```diff
@@ -81,9 +81,14 @@
+	if(user.wear_mask == src)
 		user.wear_mask_update(src, toggle_off = mask_adjusted)
-		user.update_action_buttons_icon() //when mask is adjusted out, we update all buttons icon so the user's potential internal tank correctly shows as off.
+	if(src.loc == user)
```
Ooh noted. Rookie mistake. Thank you for catching it :)
Fixed by commit ffe83e90a18854cc75d31939743d313bb50ad97c

Floofies commented on code/modules/mob/living/carbon/human/inventory.dm:
```diff
@@ -287,11 +287,49 @@
+		toggle_open_internals(tank, is_external)
+		return TRUE
+	// Use mask in absence of tube.
+	if(wear_mask && isclothing(wear_mask) && ((wear_mask.visor_flags & MASKINTERNALS) || (wear_mask.clothing_flags & MASKINTERNALS)))
```
Fixed by commit ffe83e90a18854cc75d31939743d313bb50ad97c

Floofies committed:
Fix bug relating to which tank was shown to be opened/closed

Floofies committed:
Added correct return paths on internals open/close/toggle, TRUE on success, etc.

Floofies committed:
Removed excess 'return FALSE's

Fikou commented on code/modules/mob/living/carbon/human/human.dm:
```diff
@@ -81,15 +81,13 @@
-			. += "Internal Atmosphere Info: [internal.name]"
-			. += "Tank Pressure: [internal_air.return_pressure()]"
-			. += "Distribution Pressure: [internal.distribute_pressure]"
+	if (external || internal)
+		var/obj/item/tank/target_tank = external ? external : internal
```
```suggestion
	var/obj/item/tank/target_tank = internal || external
	if(target_tank)
```

Fikou commented on code/modules/mob/living/carbon/human/inventory.dm:
```diff
@@ -287,11 +287,45 @@
+		return toggle_open_internals(tank, is_external)
+	// Invalid helmet and missing mask.
+	// Don't show the "isn't sealed" message for non-helmet headgear, such as hats.
+	if (head && (istype(head, /obj/item/clothing/head/mod) || istype(head, /obj/item/clothing/head/helmet)))
```
i dont like this check, why not just check for headwear and put this after the mask check

Floofies replied:
I thought this out quite a bit. I implemented it this way to decouple the chat notification from the actual usage of the helmet. Not all headwear makes sense with a "this isn't sealed" message, such as a flower on your head. With the check set up this way, it also adjusts the breath mask even if you have a helmet on, assuming that user wants to use their worn mask.

Floofies replied:
Fixed by commit c439d7f25b5c7423a1b6a5c242af2656ce4044d6

Fikou commented on code/modules/mob/living/carbon/life.dm:
```diff
@@ -303,21 +303,26 @@
+		// Return without taking a breath if there is no air tank.
+		return
+	// To differentiate between no internals and active, but empty internals.
+	return . ? . : FALSE
```
```suggestion
	return . || FALSE
```

Fikou commented on code/modules/mod/mod_control.dm:
```diff
@@ -444,9 +444,18 @@
+/obj/item/mod/control/proc/set_wearer(mob/living/carbon/human/user)
 	wearer = user
 	SEND_SIGNAL(src, COMSIG_MOD_WEARER_SET, wearer)
+	RegisterSignal(src, COMSIG_MOD_TOGGLED, .proc/mod_suit_toggled)
```
What's the point of registering signal to your own proc, just call it in the proc.

Floofies replied:
It's because I was advised to by some other maintainers. I can change it if you prefer

Floofies replied:
Fixed by commit 40e719a5d56e10bc5f559e8aee02340feeb574fe

Floofies committed:
Swap ternary for OR

Co-authored-by: Fikou <23585223+Fikou@users.noreply.github.com>

Floofies committed:
Simplify logic in get_status_tab_items()

Co-authored-by: Fikou <23585223+Fikou@users.noreply.github.com>

Floofies commented on code/game/objects/items/tanks/tanks.dm:
```diff
@@ -190,7 +190,7 @@
 	if(!istype(C))
 		C = loc.loc
-	if(istype(C) && C.internal == src)
+	if(istype(C) && (C.external == src || C.internal == src))
```
Can-do!

Floofies replied:
Fixed by commit ae112d9335ade9c9244a9337db2a0fd9889d8322

Floofies commented on code/game/objects/items/tanks/tanks.dm:
```diff
@@ -44,38 +44,38 @@
+
+/// Closes the tank if dropped while open.
+/obj/item/tank/dropped(mob/living/user, silent)
+	. = ..(user, silent)
```
Those procs have `SHOULD_CALL_PARENT(TRUE)`

Floofies replied:
Fixed by commit 4c0f3cfd7056557a7053a2327b1999b93be53718

Floofies commented on code/game/objects/items/tanks/tanks.dm:
```diff
@@ -44,38 +44,38 @@
+
+/// Closes the tank if given to another mob while open.
+/obj/item/tank/equipped(mob/living/user, slot, initial)
+	. = ..(user, slot, initial)
```
Those procs have `SHOULD_CALL_PARENT(TRUE)`

Floofies replied:
Fixed by commit 4c0f3cfd7056557a7053a2327b1999b93be53718

Floofies committed:
Clean up single-letter variables, src.loc usage, etc

Floofies committed:
Remove mod_suit_toggled signal handler, added internals check to mod/control/proc/seal_part

Floofies committed:
Removing unneccessary proccall arguments

Floofies committed:
Change toggle_internals to be more generic with its messages

LemonInTheDark commented on code/game/objects/items/tanks/tanks.dm:
```diff
@@ -44,37 +44,37 @@
 	/// List containing reactions happening inside our tank.
 	var/list/reaction_info
+	/// Mob that is currently breathing from the tank.
+	var/mob/living/carbon/breathing_mob = null
```
what happens if the mob is deleted?

Floofies replied:
That's a good point. I should change the tanks calls to it to `breathing_mob?.close_internals()` etc

Floofies replied:
Fixed by commit 4ecc5a90b43978f60ecb50e92a76e0fa900015ad

LemonInTheDark commented on code/game/objects/items/tanks/tanks.dm:
```diff
@@ -44,37 +44,37 @@
+	if (breathing_mob && (loc != breathing_mob))
+		breathing_mob.cutoff_internals()
 
-/obj/item/tank/proc/toggle_internals(mob/user)
-	var/mob/living/carbon/human/H = user
-	if(!istype(H))
-		return
+/// Closes the tank if given to another mob while open.
+/obj/item/tank/equipped(mob/living/user, slot, initial)
+	. = ..()
+	// Close open air tank if it was equipped by a mob other than the current user.
+	if (breathing_mob && (user != breathing_mob))
+		breathing_mob.cutoff_internals()
```
it would be nice to handle this with signals, mostly to avoid the hardref

Floofies replied:
Fixed by commit 4ecc5a90b43978f60ecb50e92a76e0fa900015ad

LemonInTheDark commented on code/modules/mob/living/carbon/life.dm:
```diff
@@ -303,21 +303,26 @@
+		cutoff_internals()
+		return
+	if (external)
+		. = external.remove_air_volume(volume_needed)
+	else if (internal)
+		. = internal.remove_air_volume(volume_needed)
```
why not just return here?

Floofies replied:
I tried early returns and it mucks the logic up. Like this:
```
if (external)
    return external.remove_air_volume(volume_needed)
else if (internal)
    return internal.remove_air_volume(volume_needed)
```
This causes a bug that allows you to breathe from your turf despite being connected to an open, but empty tank. The below conditional was definitely required. Maybe there is a simpler way to write it, but I couldn't think of one.
```
// To differentiate between no internals and active, but empty internals.
    return . || FALSE
```

Floofies committed:
Add proper handling of refs in ovj/item/tank/dropped

Floofies committed:
Cleanup comment
